modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryWithConcurrentRollbackTest.java

```diff
@@ -258,6 +264,129 @@ else if (g1Keys.contains(key))
+                cache.put(g2Key, Boolean.TRUE);
+
+                TransactionProxyImpl<?, ?> p = (TransactionProxyImpl<?, ?>)tx;
+
+                p.tx().prepare(true);
+
+                txs0 = txs(grid0);
+                txs1 = txs(grid1);
+                txs2 = txs(grid2);
+
+                assertTrue(txs0.size() == 1);
+                assertTrue(txs1.size() == 1);
+                assertTrue(txs2.size() == 1);
+
+                grid1.events().localListen(new PE() {
+                    @Override public boolean apply(Event evt) {
+                        grid1NodeLeftEventLatch.countDown();
+
+                        return true;
+                    }
+                }, EventType.EVT_NODE_LEFT);
+
+                // Block recovery procedure processing on grid1
+                grid1.context().pools().getSystemExecutorService().execute(() -> U.awaitQuiet(grid1BlockLatch));
+
+                int stripe = U.safeAbs(p.tx().xidVersion().hashCode());
+
+                stripeHolder[0] = stripe;
+
+                // Block stripe tx recovery request processing on grid1
+                grid1.context().pools().getStripedExecutorService().execute(stripe, () -> U.awaitQuiet(grid1BlockLatch));
+
+                // Prevent finish request processing on grid0
+                spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid0.name());
+
+                // Prevent finish request processing on grid1
+                spi(grid2).blockMessages(GridDhtTxFinishRequest.class, grid1.name());
+
+                runAsync(() -> {
+                    grid2.close();
+                    return null;
+                });
+            }
+            catch (Exception ignored) {
+                // Expected.
+            }
+
+            // Wait until grid1 node detects primary node left
+            grid1NodeLeftEventLatch.await();
+
+            // Wait until grid1 receives the tx recovery request and the corresponding processing task is added into the queue
+            assertTrue("tx recovery request received on grid1", GridTestUtils.waitForCondition(() ->
+                    grid1.context().pools().getStripedExecutorService().queueStripeSize(stripeHolder[0]) == 1, 5_000));
+
+            // Unblock processing in grid1. Simultaneously in striped and system pools to start
+            // recovery procedure and the tx recovery request processing at the "same" moment
+            // (for the same transaction). This should increase chances for race condition occur in
+            // the IgniteTxAdapter::markFinalizing.
+            grid1BlockLatch.countDown();
+
+            waitForTopology(2);
+
+            try {
+                awaitPartitionMapExchange();
+            }
+            catch (IgniteException ex) {
+                log.error("exception from the awaitPartitionMapExchange", ex);
+                fail("fail to wait for the partition map exchange, iteration=" + iter);
+            }
```

Comment:
Any reason to rethrow an exception?
Looks like `awaitPartitionMapExchange();` already do whan necessary.

Reply:
Just for iteration count logging - removed.